src/modules/auth/auth.service.ts

```diff
@@ -0,0 +1,107 @@
+import { Repository } from 'typeorm';
+import * as argon from 'argon2';
+import { Injectable, UnauthorizedException } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { JwtService } from '@nestjs/jwt';
+import { ConfigService } from '@nestjs/config';
+import { MailerService } from '@nestjs-modules/mailer';
+import { Doctor } from '@entities/Doctor';
+import { SentMessageInfo } from 'nodemailer';
+import { SignupDoctorDto } from './dto/signup-doctor.dto';
+import { ReconfirmDoctorDto } from './dto/reconfirm-doctor.dto';
+
+@Injectable()
+export class AuthService {
+  constructor(
+    @InjectRepository(Doctor) private doctorRepo: Repository<Doctor>,
+    private email: MailerService,
+    private jwt: JwtService,
+    private config: ConfigService,
+  ) {}
+
+  async signup(dto: SignupDoctorDto): Promise<string> {
+    const hash = await argon.hash(dto.password);
+    const token = await this.getToken({ email: dto.email });
+    const link = `${this.config.get('BASE_SERVER_URL')}/auth/confirm/${token}`;
+
+    const doctor = this.doctorRepo.create({
+      ...dto,
+      password: hash,
+      token,
+    });
+
+    await this.doctorRepo.save(doctor);
+    await this.sendConfirmationLink(dto.email, link);
+
+    return link;
+  }
+
+  async confirm(token: string): Promise<void> {
+    await this.jwt.verifyAsync(token, {
+      secret: this.config.get('JWT_SECRET'),
+    });
+
+    await this.doctorRepo
+      .createQueryBuilder('doctor')
+      .update(Doctor)
+      .set({
+        token: null,
+        isVerified: true,
+        updatedAt: new Date(),
+      })
+      .where('token = :token', { token })
+      .execute()
+      .then((response) => {
+        if (!response.affected)
+          throw new UnauthorizedException('Invalid confirmation link!');
+      });
+  }
+
+  async reconfirm(dto: ReconfirmDoctorDto): Promise<string> {
+    const token = await this.getToken({ email: dto.email });
+    const link = `${this.config.get('BASE_SERVER_URL')}/auth/confirm/${token}`;
+
+    await this.doctorRepo
+      .createQueryBuilder('doctor')
+      .update(Doctor)
+      .set({
+        token,
+        updatedAt: new Date(),
+      })
+      .where('doctor.email = :email', { email: dto.email })
+      .andWhere('doctor.is_verified = 0')
+      .execute()
+      .then((response) => {
+        if (!response.affected)
+          throw new UnauthorizedException(
+            "You don't need to verify you account.",
+          );
+      });
+
+    await this.sendConfirmationLink(dto.email, link);
+
+    return link;
+  }
+
+  async getToken(payload: string | object | Buffer): Promise<string> {
+    return this.jwt.signAsync(payload, {
+      expiresIn: this.config.get('CONFIRMATION_TOKEN_EXPIRED_AT'),
+      secret: this.config.get('JWT_SECRET'),
+    });
+  }
+
+  public sendConfirmationLink(
+    to: string,
+    link: string,
+  ): Promise<SentMessageInfo> {
+    return this.email.sendMail({
```

Comment:
better make separate files with templates for each email. in case you need find some, it will be easear then check all services, for example if you need to change text (the most common case)

Reply:
yes, and I prepared special template for sending confirmation link. His name "welcome", I didn't hardcode email text, this is name of special "pug" template.